delta-lake/common/src/main/scala/com/nvidia/spark/rapids/delta/GpuRapidsRepartitionByFilePath.scala

```diff
@@ -0,0 +1,164 @@
+import org.apache.spark.sql.types.{LongType, StringType, StructField}
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+object RapidsRepartitionByFilePath {
+  val ROW_ID_COL = "_metadata_row_id"
+  val ROW_ID_FIELD = StructField(ROW_ID_COL, LongType, nullable = false)
+  val FILE_PATH_COL = "_metadata_file_path"
+  val FILE_PATH_FIELD = StructField(FILE_PATH_COL, StringType, nullable = false)
+}
+
+case class RapidsRepartitionByFilePath(optNumPartition: Option[Int],
+    child: LogicalPlan) extends UnaryNode {
+  override def output: Seq[Attribute] = child.output
+
+  override protected def withNewChildInternal(newChild: LogicalPlan): LogicalPlan =
+    copy(child = newChild)
+}
+
+case class RapidsRepartitionByFilePathExec(optNumPartition: Option[Int],
+    child: SparkPlan) extends UnaryExecNode {
+
+  private lazy val numPartitions = optNumPartition.getOrElse(conf.numShufflePartitions)
+
+  override protected def doExecute(): RDD[InternalRow] = {
+    throw new UnsupportedOperationException("RapidsParquetFileScanExec should not be executed")
+  }
+
+  override def output: Seq[Attribute] = child.output
+
+  override protected def withNewChildInternal(newChild: SparkPlan): SparkPlan =
+    copy(child = newChild)
+
+  override def outputPartitioning: Partitioning = {
+    child.output.find(_.name == FILE_PATH_FIELD.name) match {
+      case Some(attr) => HashPartitioning(Seq(attr), numPartitions)
+      case None => throw new IllegalArgumentException(
+        s"Cannot find ${FILE_PATH_FIELD.name} in the output of the child plan")
+    }
+  }
+}
+
+case class GpuRapidsRepartitionByFilePathExec(optNumPartition: Option[Int],
+    gpuPartitionIdExpr: GpuExpression,
+    child: SparkPlan) extends UnaryExecNode with GpuExec {
+  private lazy val numPartitions = optNumPartition.getOrElse(conf.numShufflePartitions)
+
+  override def output: Seq[Attribute] = child.output
+
+  override def outputPartitioning: Partitioning = {
+    child.output.find(_.name == FILE_PATH_FIELD.name) match {
+      case Some(attr) => HashPartitioning(Seq(attr), numPartitions)
+      case None => throw new IllegalArgumentException(
+        s"Cannot find ${FILE_PATH_FIELD.name} in the output of the child plan")
+    }
+  }
+
+  // The same as what feeds us
+  override def outputBatching: CoalesceGoal = GpuExec.outputBatching(child)
+
+  protected override def doExecute(): RDD[InternalRow] = throw new UnsupportedOperationException(
+    s"${getClass.getCanonicalName} does not support row-based execution")
+
+  override protected def internalDoExecuteColumnar(): RDD[ColumnarBatch] = {
+    val rdd = child.executeColumnar()
+    if (numPartitions == 1 && rdd.getNumPartitions < 1) {
+      // Make sure we don't output an RDD with 0 partitions, when claiming that we have a
+      // `SinglePartition`.
+      new GpuCoalesceExec.EmptyRDDWithPartitions(sparkContext, numPartitions)
+    } else {
+      rdd.coalesce(numPartitions,
```

Comment:
What if numPartitions > rdd.getNumPartitions?  Without a shuffle, this can't spread to a wider number of partitions.